rust/datafusion/src/execution/physical_plan/hash.rs

```diff
@@ -0,0 +1,176 @@
+/// for floating point numerics)
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum KeyScalar {
+    /// Boolean
+    Boolean(bool),
+    /// 8 bits
+    UInt8(u8),
+    /// 16 bits
+    UInt16(u16),
+    /// 32 bits
+    UInt32(u32),
+    /// 64 bits
+    UInt64(u64),
+    /// 8 bits signed
+    Int8(i8),
+    /// 16 bits signed
+    Int16(i16),
+    /// 32 bits signed
+    Int32(i32),
+    /// 64 bits signed
+    Int64(i64),
+    /// string
+    Utf8(String),
+}
+
+/// Return an KeyScalar from an ArrayRef of a given row.
+pub fn create_key(col: &ArrayRef, row: usize) -> Result<KeyScalar> {
+    match col.data_type() {
+        DataType::Boolean => {
+            let array = col.as_any().downcast_ref::<BooleanArray>().unwrap();
+            return Ok(KeyScalar::Boolean(array.value(row)));
+        }
+        DataType::UInt8 => {
+            let array = col.as_any().downcast_ref::<UInt8Array>().unwrap();
+            return Ok(KeyScalar::UInt8(array.value(row)));
+        }
+        DataType::UInt16 => {
+            let array = col.as_any().downcast_ref::<UInt16Array>().unwrap();
+            return Ok(KeyScalar::UInt16(array.value(row)));
+        }
+        DataType::UInt32 => {
+            let array = col.as_any().downcast_ref::<UInt32Array>().unwrap();
+            return Ok(KeyScalar::UInt32(array.value(row)));
+        }
+        DataType::UInt64 => {
+            let array = col.as_any().downcast_ref::<UInt64Array>().unwrap();
+            return Ok(KeyScalar::UInt64(array.value(row)));
+        }
+        DataType::Int8 => {
+            let array = col.as_any().downcast_ref::<Int8Array>().unwrap();
+            return Ok(KeyScalar::Int8(array.value(row)));
+        }
+        DataType::Int16 => {
+            let array = col.as_any().downcast_ref::<Int16Array>().unwrap();
+            return Ok(KeyScalar::Int16(array.value(row)));
+        }
+        DataType::Int32 => {
+            let array = col.as_any().downcast_ref::<Int32Array>().unwrap();
+            return Ok(KeyScalar::Int32(array.value(row)));
+        }
+        DataType::Int64 => {
+            let array = col.as_any().downcast_ref::<Int64Array>().unwrap();
+            return Ok(KeyScalar::Int64(array.value(row)));
+        }
+        DataType::Utf8 => {
+            let array = col.as_any().downcast_ref::<StringArray>().unwrap();
+            return Ok(KeyScalar::Utf8(String::from(array.value(row))));
+        }
+        _ => {
+            return Err(ExecutionError::ExecutionError(
+                "Unsupported key data type".to_string(),
+            ))
+        }
+    }
+}
+
+/// Create array from `key` attribute in map entry (representing a grouping scalar value)
+macro_rules! key_array_from_map_entries {
+    ($BUILDER:ident, $TY:ident, $MAP:expr, $COL_INDEX:expr) => {{
+        let mut builder = $BUILDER::new($MAP.len());
+        let mut err = false;
+        for k in $MAP.keys() {
```

Comment:
Side note: The current impl walks the `keys` and `values` from the map separately when building the final result. It would perhaps be safer and more efficient to walk the entries instead (as a separate PR).